vault/ssorolecredentialsprovider.go

```diff
@@ -122,7 +129,19 @@ func (p *SSORoleCredentialsProvider) getOIDCToken(ctx context.Context) (token *s
 			return token, true, nil
 		}
 	}
-	token, err = p.newOIDCToken(ctx)
+
+	// if we must use stdout (either by user choice or because we have determined we are in an SSH session where we cannot open a browser)
+	// then we use the "device code" grant flow, and print URLs to stdout for the user to manually copy/paste into their browser.
+	//
+	// Otherwise we use the "authorization code" grant flow with Proof Key for
+	// Code Exchange (PKCE) and open the browser automatically. The latter flow
+	// is more user-friendly and secure because the step comparing the challenge
+	// code between the CLI and browser is automated entirely.
+	if p.UseStdout {
```

Comment:
I commonly use `--stdout` even when I'm on the same machine as where `aws-vault` is running, for various purposes (e.g. wanting to login in a non-default browser). I'm not 100% convinced that `UseStdout` should be the basis for chosing between the device code and PKCE flow, how about instead following the same behavior as the AWS CLI? It seems to have a flag `--use-device-code` flag https://github.com/aws/aws-cli/commit/130005af5ea6a75705ed528aaf06d533f449bef9#diff-e07a10a6eb1a677e905b0498651e672137b5dff19d357933bd7bc3e36f845a3bL42 which defaults to false

Reply:
Great shout. I hadn't considered the same device non-default browser path that users might wish to take. I was primarily thinking about remote workstations and in that context printing the URL out to be opened locally would not make sense as the eventual redirect destination would be inaccessible to the browser.